Fix ProjectHandle input

aeolianeth opened the pull request:
Fixes #119, https://github.com/jbx-protocol/juice-interface/issues/189

When editing a payout, the project handle field wasn't populating automatically.

This PR fixes this by:
- Leveraging `Form.Item`'s automatic data binding correctly
- Adding a loading spinner to the project handle input

There's a couple of other unrelated fixes in here, too. I've made comments throughout the PR to highlight these 🙂 .

### Demo

https://user-images.githubusercontent.com/94939382/143811934-5942f732-3a38-43e7-91b3-81efb8f637f8.mp4

aeolianeth commented on src/components/Create/PayModsForm.tsx:
```diff
@@ -51,7 +51,6 @@ export default function PayModsForm({
 
       <Form layout="vertical">
         <FormItems.ProjectPayoutMods
-          name="mods"
```
This doesn't need a name. Removing it fixes a console error:

```
Warning: [antd: Form.Item] `children` is array of render props cannot have `name`. 
```

<img width="658" alt="Screen Shot 2021-11-29 at 2 40 07 pm" src="https://user-images.githubusercontent.com/94939382/143812273-e3dcf10f-1b85-4be1-bce3-bc6d25eea1eb.png">

aeolianeth replied:
This doesn't need a name. Removing it fixes a console error:

```
Warning: [antd: Form.Item] `children` is array of render props cannot have `name`. 
```

<img width="658" alt="Screen Shot 2021-11-29 at 2 40 07 pm" src="https://user-images.githubusercontent.com/94939382/143812273-e3dcf10f-1b85-4be1-bce3-bc6d25eea1eb.png">

aeolianeth commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -80,16 +89,31 @@ export default function ProjectHandle({
-  if (handleExists && requireState === 'exists') {
-    suffix = <CheckCircleOutlined style={{ color: colors.icon.success }} />
-  }
+  const suffix: string | JSX.Element = useMemo(() => {
```
Changed `suffix` to memoized value. I _think_ this is a good use case, please correct me if not.

peripheralist replied:
my understanding is that memoization is kinda subjective—just meant to be used to prevent expensive logic (i.e. network calls or complex calculations) from being run unnecessarily. i'm down to use it here

aeolianeth commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -114,12 +138,12 @@ export default function ProjectHandle({
           setInputContents(val)
-
-          if (returnValue !== 'id') {
-            onValueChange(val)
```
The parent `Form.Item` is already capturing/handling input value changes.

aeolianeth commented on src/components/shared/formItems/ProjectPayoutMods.tsx:
```diff
@@ -419,14 +417,13 @@ export default function ProjectPayoutMods({
           ) : (
             <FormItems.ProjectHandle
-              name="project"
+              name="handle"
```
Setting the `name` to `handle` to match the form state variable ensures that the `Form.Item` binds the `input` to the `form` state correctly.

aeolianeth commented on src/components/shared/formItems/ProjectPayoutMods.tsx:
```diff
@@ -419,14 +417,13 @@ export default function ProjectPayoutMods({
                 label: 'Project handle',
                 rules: [{ required: true }],
               }}
-              value={form.getFieldValue('handle')}
```
By setting the `name` correctly above, we don't need to manually retrieve and set the form field 🙂 .

aeolianeth commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -17,29 +17,32 @@ export default function ProjectHandle({
   } = useContext(ThemeContext)
   const [inputContents, setInputContents] = useState<string>()
-
+  const [handleLoading, setHandleLoading] = useState<boolean>(false)
```
Adding a loading state lets us display a spinner in the input field.

Ideally we would debounce the input so we don't re-render every input type. Perhaps this can be a follow-up issue.

peripheralist replied:
debouncing would be great here, lots of unnecessary calls right now. we can make that a separate issue

aeolianeth replied:
https://github.com/jbx-protocol/juice-interface/issues/243 👍

aeolianeth commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -17,29 +17,32 @@ export default function ProjectHandle({
-      setInputContents(value)
-    } else if (isBigNumberish(value)) {
+    if (initialValue) {
+      setHandleLoading(true)
```
If initialValue is truthy, display the spinner. `idForHandle`'s callback will remove the spinner in its callback fn.

aeolianeth commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -17,29 +17,32 @@ export default function ProjectHandle({
   formItemProps,
   onValueChange,
-  value,
+  initialValue,
```
I thought `initialValue` was clearer than `value`. Loosely held opinion, though.

aeolianeth commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -54,21 +57,27 @@ export default function ProjectHandle({
-        : () => null,
-      [returnValue],
+      id => {
+        setHandleLoading(false)
```
Remove the loading spinner once we get an `id` for the handle.

`suffix` (below) will display the correct help text in the input field.

aeolianeth commented on src/components/shared/formItems/ProjectPayoutMods.tsx:
```diff
@@ -41,7 +41,6 @@ type ModType = 'project' | 'address'
 type EditingPayoutMod = PayoutMod & { handle?: string }
 
 export default function ProjectPayoutMods({
-  name,
```
See https://github.com/jbx-protocol/juice-interface/pull/241/files#r758043049

aeolianeth commented:
Yo @peripheralist, another weird Antd form issue fix  - could you review when you have a chance?

I did my best with local testing, but would appreciate if you could test the other uses of `ProjectHandle`, too 🙏 .

Thanks!

peripheralist commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -54,21 +57,27 @@ export default function ProjectHandle({
   })
 
-  const handleExists = idForHandle?.gt(0)
+  const handleExists = Boolean(idForHandle?.gt(0))
```
what's the purpose here? does `Boolean(x)` convert all falsy values to `false`?

aeolianeth replied:
> does Boolean(x) convert all falsy values to false?

Exactly 👍

peripheralist commented on src/components/shared/formItems/ProjectHandle.tsx:
```diff
@@ -114,12 +138,12 @@ export default function ProjectHandle({
         placeholder="handle"
         type="string"
         autoComplete="off"
+        spellCheck={false}
```
👍

aeolianeth commented:
Copying comment from issue (https://github.com/jbx-protocol/juice-interface/issues/119#issuecomment-982075205)

> This is looking great, but I'm noticing the issue still remains when editing a payoutMod—the initial value doesn't populate. Looks like the initialValue is undefined when it shouldn't be, so probably not a bug in your code.
> 
> After poking around a bit, it seems like an issue with how the initialValue is being passed from ProjectPayoutMods. If you wanna take a look at it you might be able to work in a fix with this PR, otherwise I'll start working on it separately later tonight

peripheralist commented:
> Copying comment from issue ([#119 (comment)](https://github.com/jbx-protocol/juice-interface/issues/119#issuecomment-982075205))

glad you found this! put it in the wrong place obviously.

peripheralist commented:
@aeolianeth try rebasing this against main. just pushed a patch that now makes sure the projectId is passed to the projectHandle form item.

jnoh commented:
Linking: https://github.com/jbx-protocol/juice-interface/issues/189

aeolianeth commented:
@jnoh thanks for much for identifying the fix to https://github.com/jbx-protocol/juice-interface/issues/189. I took inspiration from your PR and added a bit of extra logic so we show the green tick when a project doesn't change their handle.

Let me know what you think!

https://github.com/jbx-protocol/juice-interface/pull/241/commits/d91fcbe172ba5af82d44fb6a4aa00a105a452b0d

aeolianeth commented:
@peripheralist thanks for the projectId fix! Rebased and good to go.

I also added an extra commit https://github.com/jbx-protocol/juice-interface/commit/d91fcbe172ba5af82d44fb6a4aa00a105a452b0d thanks to @jnoh's insights (to fix https://github.com/jbx-protocol/juice-interface/issues/189)🤝

peripheralist commented:
sweet! would be great to solve both these in this PR. playing with this i still don't see the handle populate, though guessing the value is there since the suffix shows a green check mark. haven't looked into the code 
![image](https://user-images.githubusercontent.com/79433522/144116382-469de246-fe02-4898-b893-8da7f4869dc6.png)

peripheralist commented:
also just a heads up, a lot of Fleek builds have been failing lately due to what looks like something on their end. hoping that gets resolved soon

aeolianeth commented:
@peripheralist sorry about that! Think I missed something during the rebase 🙈 .

99% sure everything is gucci now. Here's 2 videos to demonstrate:

### Fix for #119:

https://user-images.githubusercontent.com/94939382/144166927-3a8288cf-27d2-4407-a448-c4700f6b5c4c.mp4

### Fix for #189:

https://user-images.githubusercontent.com/94939382/144167102-0839b1b9-9fa4-4997-b56a-32cba42eab95.mp4

Let me know what you think 🙏

peripheralist commented:
seems like there are some deeper bugs here. seems like this is working fine from the create form, but it doesn't work when editing existing payouts. 

maybe related to how the projectID is being passed to the ProjectPayoutMods form?

For details and discussion space check out payout-errorgardener-fc22 

https://user-images.githubusercontent.com/79433522/144299383-a20fd465-a34f-45f6-bba7-b80469b6e189.mov

aeolianeth commented:
@peripheralist ah I see! Thanks for the demo video 🙏 

This one is turning out to be a bit of a doozy. I've almost fixed everything in https://github.com/jbx-protocol/juice-interface/pull/241/commits/7b17e19e01c8ece78fe5868cf03b2db1a3a4bb00, however there's still an issue where the `useContractReader` is triggered too many times, causing the loading state to appear too frequently.

I'll keep working on this, but if you have a chance, please feel free to have a play and see if you can find a quick fix.

aeolianeth commented:
@peripheralist happy to say that I think I've finally figured this out.

Because of the "magic" binding that `Antd`'s `Form.Item` does, the solution was to use a custom input component ([kinda described in their docs](https://ant.design/components/form/#components-form-demo-customized-form-controls)). This approach allows us to "break" the binding and create a unidirectional data flow from `input` _up_ to `form.item`; our custom input doesn't accept a value prop, so `form.item` can't influence it.

There's a few other minor improvements here and there, and I've identified a few others that I'll put in seperate PRs.

Note that I didnt test `TokenRefs` on the Edit Tracked Assets modal as I didn't know how 🙈 .

Let me know what you think 🙏 

### Create project
https://user-images.githubusercontent.com/94939382/144680592-ff5d868e-d3f0-49b2-bf8f-82ebfdb8e33d.mp4

### Edit Project
https://user-images.githubusercontent.com/94939382/144680641-33fbbea5-64d0-4ad0-a82f-6693c4c01a7e.mp4

### Edit Payouts
https://user-images.githubusercontent.com/94939382/144680696-98add718-e911-4336-90fb-a1d94bf33084.mp4

cc @jnoh

peripheralist reviewed (approved):
nice detective work! that custom input `value` behavior is helpful to know. 

this is working great, thanks for spending so much time on this 🙏
